lib/vrl/stdlib/src/encrypt.rs

```diff
@@ -0,0 +1,343 @@
+use aes::cipher::block_padding::{AnsiX923, Iso10126, Iso7816, Pkcs7};
+use aes::cipher::generic_array::GenericArray;
+use aes::cipher::{AsyncStreamCipher, BlockEncryptMut};
+use vrl::prelude::*;
+
+use aes::cipher::KeyIvInit;
+use aes::cipher::StreamCipher;
+
+type Aes128Cbc = cbc::Encryptor<aes::Aes128>;
+type Aes192Cbc = cbc::Encryptor<aes::Aes192>;
+type Aes256Cbc = cbc::Encryptor<aes::Aes256>;
+
+pub(crate) fn get_key_bytes<const N: usize>(key: Value) -> Result<[u8; N]> {
+    let bytes = key.try_bytes()?;
+    if bytes.len() != N {
```

Comment:
think the typical use case would be a customer has a known key on a different server already (where they will do the decrypting), and it is also used in Vector (to encrypt). That means the actual key is likely loaded from the environment (env vars are currently supported, and other methods might be added). I'm not sure  understand the use case where a key would be generated in Vector, since that key would need to be sent to another system for it to be able to decrypt

Reply:
Maybe I need to clarify that the key here is expected to be bytes, not plaintext. It just uses a `string` in VRL (which stores a `Bytes` internally). A typicaly use case could be an encryption key is stored in an env variable base64 encoded, and it is decoded in Vector (which returns it as a string)

Reply:
For clarification, the algorithms I listed do not _create_ the key, they are hash algorithm that turn a plain text key of variable length into a fixed-length binary, with the modifications like possible addition of a salt and extra rounds of hashing to make reversing the process (ie with rainbow tables) infeasible. Most systems make the separating this stage out difficult, as the internal key is never exposed.

This actually prompts a larger question about pre-computation. Just like with regex searches, symmetric encryption can conceptually be broken up into two stages where the first is independent of the input data. This includes both the hashing of the key phrase into a final key data, and then AES has a stage that prepares internal data based on the key bits. IIRC VRL has built-in support for pre-compiling regexes, but we might want to consider making that facility more generic to cover use cases like this as well.

Reply:
Fair enough. I suspect that we will add more functions in the future for obtaining a key, which might include one of these suggestions. I agree it's out of scope for now though.

Reply:
The data I was referencing would be the key, which would be known at startup but wouldn't necessarily be a literal in the script.

However, @JeanMertz is right, the numbers are more important than speculation. Do these interfaces have a good way of precomputing the key setup?

Reply:
:+1: to getting some benchmark numbers. The VRL benchmark suite should illuminate.